Add String.replace

thiagoarrais opened the pull request:
Work in progress pull request for fixing #116 

The implementation isn't complete yet, but I intend to incorporate changes based on feedback. For starters, just let me know if this is the structure you expect. I'll try to evolve the `replace` method to be spec-compliant.

Related question: does this project use tests? Where should I add them if so?

TODO List:

- [x] Do not interpret string pattern as regex
- [x] Support regex pattern
- [x] Support funtion replacement
- [x] ~~Include function length (related: #193)~~

jasonwilliams commented:
> Related question: does this project use tests? Where should I add them if so?
Yes, there are quite a few tests in this project, each file (or module) has the tests embedded with the code. So for String, you will see the tests at the bottom of the file. Im sure there are enough tests here for you to familiar yourself with.
https://github.com/jasonwilliams/boa/blob/master/src/lib/builtins/string.rs#L781-L1044

jasonwilliams commented:
Im not sure if using regex directly will work, as the rules may differ from rust to JS.
We have a RegExp object which has `match` here:
https://github.com/jasonwilliams/boa/blob/master/src/lib/builtins/regexp.rs#L257-L274

Maybe you could use that?

Secondly incase you're unaware your rustfmt is failing:
https://github.com/jasonwilliams/boa/pull/190/checks?check_run_id=273620005

You can run `cargo fmt --verbose` in the project

thiagoarrais commented:
Nice, @jasonwilliams! I'll add tests (sorry, hadn't located them just out of Rust-noobness) and rustfmt sometime today.

About RegExp, I started with that, but Rust complained about it not being exported or something. But my final version will most likely use that.

Thanks for the feedback!

thiagoarrais commented:
This is taking some shape now.

I have included a `get_argument` function inspired by but different than [`regexp::get_argument`](https://github.com/jasonwilliams/boa/blob/master/src/lib/builtins/regexp.rs#L40). It is designed to mimic JS' behaviour of interpreting missing parameters as `undefined`. I suspect such a function already exists somewhere in the codebase. Care to help me find it?

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -766,6 +798,7 @@ pub fn create_constructor(global: &Value) -> Value {
     proto.set_field_slice("substr", to_value(substr as NativeFunctionData));
     proto.set_field_slice("valueOf", to_value(value_of as NativeFunctionData));
     proto.set_field_slice("matchAll", to_value(match_all as NativeFunctionData));
+    proto.set_field_slice("replace", to_value(replace as NativeFunctionData));
```
You need to set the [`length`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/length) property of the function as well.

thiagoarrais replied:
I don't know if I understood that. Do you mean `str.replace.length` should return `2`? If so, would you say that `str.charAt.length` should return `1` then?

IovoslavIovchev replied:
Yes, to both your questions. A `function`'s length represents the number of "formal" parameters, e.g. those with no default value.

thiagoarrais replied:
Do the others functions in `String` already support that in boa or are you suggesting that they should? I'm locally getting `str.charAt.length == undefined`, but I may have an outdated branch...

IovoslavIovchev replied:
Unfortunately, some functions in boa don't have their length set because it is easily missable.

I have suggested an approach in #193 and would very much like we have something like that, so we don't miss like things like that.

thiagoarrais replied:
Nice. I'm working on other stuff here, but I'll make sure to check where that PR of yours goes.

thiagoarrais commented on src/lib/builtins/regexp.rs:
```diff
@@ -14,9 +14,9 @@ use crate::{
+pub struct RegExp {
     /// Regex matcher.
-    matcher: Regex,
+    pub matcher: Regex,
```
Need some help here. I'm feeling a bit uneasy about publishing this. What is the idiomatic way to access regex funcionality from outside the regexp module?

IovoslavIovchev replied:
Do you intend to modify the value in any way?

The most "popular" and obvious way would be to have an immutable getter (and a mutable one only if _really_ needed).

thiagoarrais replied:
I don't intend to modify the value. I was considering getter vs. something like adding a `replace_all` delegating method. The second option could be used to deal with differences between JS and Rust regex engines.

IovoslavIovchev replied:
What exactly is your idea for `replace_all`?

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -728,6 +728,41 @@ pub fn match_all(this: &Value, args: &[Value], ctx: &mut Interpreter) -> ResultV
+    // We then do the same thing for the other args
+    // TODO: what if there are missing arguments?
+    let pattern = get_argument(args, 0);
+    if undefined() == pattern {
```
If I am not mistaken, there was an `is_undefined` method.
So you can try `pattern.is_undefined()` instead.

thiagoarrais replied:
Good catch!

thiagoarrais commented:
I think this is ready for another round of review.

My main concern here is about @IovoslavIovchev's suggestion on `length`. It doesn't seem to work, even after adapting for #206. I'm assuming this is unrelated to this change because some other functions in `String` don't return a valid length. But I'm open to adapting the code if needed.

thiagoarrais commented:
The problem with `length` seems related to #210. This should be good to go. Please let me know if you need anything else.

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -728,6 +729,60 @@ pub fn match_all(this: &Value, args: &[Value], ctx: &mut Interpreter) -> ResultV
+    // Make a Rust Regex from the first argument
+    let pattern = get_argument(args, 0);
+    if pattern.is_undefined() {
+        return Ok(to_value(this_str));
```
```suggestion
        return Ok(this.clone());
```

IovoslavIovchev replied:
This is just a small optimisation to avoid calling `ctx.value_to_rust_string` if the regex is `undefined`

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -728,6 +729,60 @@ pub fn match_all(this: &Value, args: &[Value], ctx: &mut Interpreter) -> ResultV
+pub fn replace(this: &Value, args: &[Value], ctx: &mut Interpreter) -> ResultValue {
+    // First we get it the actual string a private field stored on the object only the engine has access to.
+    // Then we convert it into a Rust String
+    let this_str: &str = &ctx.value_to_rust_string(this);
```
Move this after the `if` statement below.

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -728,6 +729,60 @@ pub fn match_all(this: &Value, args: &[Value], ctx: &mut Interpreter) -> ResultV
+                .call(&replacement, &undefined(), args)
+                .unwrap_or_else(|_| undefined());
+            ctx.value_to_rust_string(&rs_value)
+        }) as Box<dyn FnMut((&Captures)) -> (String)>
```
Remove the parentheses, so it becomes

```suggestion
        }) as Box<dyn FnMut(&Captures) -> String>
```

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -728,6 +729,60 @@ pub fn match_all(this: &Value, args: &[Value], ctx: &mut Interpreter) -> ResultV
+    Ok(to_value(result_str.into_owned()))
+}
+
+fn get_argument(args: &[Value], idx: usize) -> Value {
```
This is a bit more general, so it would be better you move it somewhere, where it can be widely accessed.

I am thinking _exec.rs_. What do you think?

thiagoarrais replied:
I actually wonder if this _already_ exists somewhere. But I can't find it...

IovoslavIovchev replied:
I know something similar exists in _regexp.rs_ but nothing else.

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -1041,4 +1097,41 @@ mod tests {
+        "#;
+
+        forward(&mut engine, init);
+        assert_eq!(forward(&mut engine, "result1"), "The quick brown fox jumps over the lazy monkey. If the monkey reacted, was it really lazy?");
```
```suggestion
        assert_eq!(forward(&mut engine, "result1"), "The quick brown fox jumps over the lazy monkey. If the dog reacted, was it really lazy?");
```

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -1041,4 +1097,41 @@ mod tests {
+        forward(&mut engine, init);
+        assert_eq!(forward(&mut engine, "result1"), "The quick brown fox jumps over the lazy monkey. If the monkey reacted, was it really lazy?");
+        assert_eq!(forward(&mut engine, "result2"), "The quick brown fox jumps over the lazy monkey. If the dog reacted, was it really lazy?");
+        assert_eq!(forward(&mut engine, "regexResult"), "The quick brown fox jumps over the lazy monkey. If the monkey.reacted, was it really lazy?");
```
```suggestion
        assert_eq!(forward(&mut engine, "regexResult"), "The quick brown fox jumps over the lazy monkey. If the dog reacted, was it really lazy?");
```

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -1041,4 +1097,41 @@ mod tests {
+        assert_eq!(forward(&mut engine, "regexResult"), "The quick brown fox jumps over the lazy monkey. If the monkey.reacted, was it really lazy?");
+        assert_eq!(
+            forward(&mut engine, "funResult1"),
+            "The quick brown fox jumps over the lazy DOG. If the DOG reacted, was it really lazy?"
```
```suggestion
            "The quick brown fox jumps over the lazy DOG. If the dog reacted, was it really lazy?"
```

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -1041,4 +1097,41 @@ mod tests {
+        );
+        assert_eq!(
+            forward(&mut engine, "funResult2"),
+            "The quick brown fox jumps over the lazy DOG. If the DOG reacted, was it really lazy?"
```
```suggestion
            "The quick brown fox jumps over the lazy DOG. If the dog reacted, was it really lazy?"
```

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -1041,4 +1097,41 @@ mod tests {
+            forward(&mut engine, "exceptional1"),
+            "The quick brown fox jumps over the lazy dog. If the dog reacted, was it really lazy?"
+        );
+        assert_eq!(forward(&mut engine, "exceptional2"), "The quick brown fox jumps over the lazy undefined. If the undefined reacted, was it really lazy?");
```
```suggestion
        assert_eq!(forward(&mut engine, "exceptional2"), "The quick brown fox jumps over the lazy undefined. If the dog reacted, was it really lazy?");
```

IovoslavIovchev commented on src/lib/builtins/string.rs:
```diff
@@ -1041,4 +1097,41 @@ mod tests {
+            "The quick brown fox jumps over the lazy dog. If the dog reacted, was it really lazy?"
+        );
+        assert_eq!(forward(&mut engine, "exceptional2"), "The quick brown fox jumps over the lazy undefined. If the undefined reacted, was it really lazy?");
+        assert_eq!(forward(&mut engine, "exceptional2"), "The quick brown fox jumps over the lazy undefined. If the undefined reacted, was it really lazy?");
```
Remove this line, it's a duplicate

IovoslavIovchev commented:
The current implementation replaces all occurrences of the first argument when it's a `String`, when, in fact, it should only replace the first occurrence. See [the spec](https://tc39.es/ecma262/#sec-string.prototype.replace) for more info.

IovoslavIovchev commented:
@thiagoarrais do you need any help with this?

thiagoarrais commented:
I think I've addressed most of your comments, @IovoslavIovchev . The one thing I couldn't tackle yet is the spec breakage regarding replacing only the first occurrence. I plan on getting back to this some time, but I'm a little short on time. If you could help on that, it would be greatly appreciated.

Maybe you can give some general direction on a good approach? I'd also include a commit of yours here if you're willing to branch from this PR.

IovoslavIovchev commented:
@thiagoarrais no worries. I will take a look.

jasonwilliams commented:
This has now been added
